File: internal/linkdb/linkdb.go
package linkdb

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
)

type DB struct {
	dir string
	mu  sync.Mutex
}

func New(dir string) *DB {
	os.MkdirAll(dir, 0755)
	return &DB{dir: dir}
}

func (db *DB) path(sub string) string {
	return filepath.Join(db.dir, sub+".json")
}

func (db *DB) Get(sub string) ([]string, error) {
	db.mu.Lock()
	defer db.mu.Unlock()
	data, err := os.ReadFile(db.path(sub))
	if os.IsNotExist(err) {
		return []string{}, nil
	}
	if err != nil {
		return nil, err
	}
	var links []string
	if err := json.Unmarshal(data, &links); err != nil {
		return nil, err
	}
	return links, nil
}

func (db *DB) save(sub string, links []string) error {
	data, err := json.Marshal(links)
	if err != nil {
		return err
	}
	return os.WriteFile(db.path(sub), data, 0600)
}

func contains(list []string, item string) bool {
	for _, v := range list {
		if v == item {
			return true
		}
	}
	return false
}

func remove(list []string, item string) []string {
	res := make([]string, 0, len(list))
	for _, v := range list {
		if v != item {
			res = append(res, v)
		}
	}
	return res
}

func (db *DB) AddLink(a, b string) error {
	linksA, err := db.Get(a)
	if err != nil {
		return err
	}
	if !contains(linksA, b) {
		linksA = append(linksA, b)
		if err := db.save(a, linksA); err != nil {
			return err
		}
	}
	linksB, err := db.Get(b)
	if err != nil {
		return err
	}
	if !contains(linksB, a) {
		linksB = append(linksB, a)
		if err := db.save(b, linksB); err != nil {
			return err
		}
	}
	return nil
}

func (db *DB) RemoveLink(a, b string) error {
	linksA, err := db.Get(a)
	if err != nil {
		return err
	}
	if contains(linksA, b) {
		if err := db.save(a, remove(linksA, b)); err != nil {
			return err
		}
	}
	linksB, err := db.Get(b)
	if err != nil {
		return err
	}
	if contains(linksB, a) {
		if err := db.save(b, remove(linksB, a)); err != nil {
			return err
		}
	}
	return nil
}

File: internal/linkdb/LinkDB.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace Tinylinks.Linkdb
{
    // Simple file-based link database
    public class LinkDB
    {
        private readonly string dir;
        private readonly object mu = new object();

        public LinkDB(string directory)
        {
            Directory.CreateDirectory(directory);
            dir = directory;
        }

        private string PathFor(string sub)
        {
            return Path.Combine(dir, sub + ".json");
        }

        public string[] Get(string sub)
        {
            lock (mu)
            {
                var path = PathFor(sub);
                if (!File.Exists(path))
                {
                    return Array.Empty<string>();
                }
                var data = File.ReadAllText(path);
                try
                {
                    var links = JsonSerializer.Deserialize<string[]>(data);
                    return links ?? Array.Empty<string>();
                }
                catch
                {
                    return Array.Empty<string>();
                }
            }
        }

        private void Save(string sub, string[] links)
        {
            var data = JsonSerializer.Serialize(links);
            File.WriteAllText(PathFor(sub), data);
        }

        private static bool Contains(string[] list, string item)
        {
            for (int i = 0; i < list.Length; i++)
            {
                if (list[i] == item) return true;
            }
            return false;
        }

        private static string[] Remove(string[] list, string item)
        {
            var res = new List<string>();
            for (int i = 0; i < list.Length; i++)
            {
                if (list[i] != item)
                {
                    res.Add(list[i]);
                }
            }
            return res.ToArray();
        }

        public void AddLink(string a, string b)
        {
            lock (mu)
            {
                var linksA = Get(a);
                if (!Contains(linksA, b))
                {
                    var newList = new string[linksA.Length + 1];
                    for (int i = 0; i < linksA.Length; i++) newList[i] = linksA[i];
                    newList[linksA.Length] = b;
                    Save(a, newList);
                }
                var linksB = Get(b);
                if (!Contains(linksB, a))
                {
                    var newList = new string[linksB.Length + 1];
                    for (int i = 0; i < linksB.Length; i++) newList[i] = linksB[i];
                    newList[linksB.Length] = a;
                    Save(b, newList);
                }
            }
        }

        public void RemoveLink(string a, string b)
        {
            lock (mu)
            {
                var linksA = Get(a);
                if (Contains(linksA, b))
                {
                    Save(a, Remove(linksA, b));
                }
                var linksB = Get(b);
                if (Contains(linksB, a))
                {
                    Save(b, Remove(linksB, a));
                }
            }
        }
    }
}
